source/dofs/dof_tools_sparsity.cc

```diff
@@ -1029,39 +1029,36 @@ namespace DoFTools
 
                   const bool periodic_neighbor = cell->has_periodic_neighbor (face);
 
-                  if (cell->at_boundary (face) && (!periodic_neighbor))
+                  for (unsigned int i=0; i<cell->get_fe().dofs_per_cell; ++i)
                     {
-                      for (unsigned int i=0; i<cell->get_fe().dofs_per_cell; ++i)
+                      const unsigned int ii
+                        = (cell->get_fe().is_primitive(i) ?
+                           cell->get_fe().system_to_component_index(i).first
+                           :
+                           cell->get_fe().get_nonzero_components(i).first_selected_component()
+                          );
+
+                      Assert (ii < cell->get_fe().n_components(), ExcInternalError());
+
+                      for (unsigned int j=0; j<cell->get_fe().dofs_per_cell; ++j)
```

Comment:
I think if I read this right, then this loop over `j` here is used for coupling DoFs on the cell itself. At least it couples DoFs on the current cell with itself in lines 1057-1058. If so, it could be pulled out of the loop over all faces, and that would probably make the code more readable as well. 

Maybe add a text like "couple dofs on the cell with each other".

Reply:
But on second thought, do you actually need this code? Isn't this already done in line 1018?

Reply:
It is also dubious because the code in lines 1057-1058 uses `flux_mask` instead of `int_mask`.

Reply:
I guess the idea was to just consider degrees of freedom that are nonzero on a face and couple them here. Of course, this still requires (to work reliably) to consider all degrees of freedom on the cell.
I think it makes sense indeed to simplify this to couple degrees of freedom on a cell if `int_mask` or `flux_mask` says so.

Reply:
Oh, I see what you mean.

But the code could still be moved out of the loop over all faces, and one then would only have to ask `if int_mask OR flux_mask`. I think that would also make the code easier to read.